test/unittest/unittest_nntrainer_models.cpp

```diff
@@ -0,0 +1,351 @@
+     << iteration;
+  std::string err_msg = ss.str();
+
+  verify(*in, expected_input, err_msg + " at input ");
+  verifyWeight(err_msg);
+  nntrainer::sharedConstTensor out = node->forwarding(in);
+  verifyWeight(err_msg);
+  verify(*out, expected_output, err_msg + " at output ");
+  return out;
+}
+
+nntrainer::sharedConstTensor
+NodeWatcher::lossForward(nntrainer::sharedConstTensor pred,
+                         nntrainer::sharedConstTensor answer, int iteration) {
+  std::stringstream ss;
+  ss << "loss failed at " << node->getName() << " at iteration " << iteration;
+  std::string err_msg = ss.str();
+
+  nntrainer::sharedConstTensor out =
+    std::static_pointer_cast<nntrainer::LossLayer>(node)->forwarding(pred,
+                                                                     answer);
+
+  return out;
+}
+
+nntrainer::sharedConstTensor
+NodeWatcher::backward(nntrainer::sharedConstTensor deriv, int iteration,
+                      bool should_verify) {
+  std::stringstream ss;
+  ss << "backward failed at " << node->getName() << " at iteration "
+     << iteration;
+  std::string err_msg = ss.str();
+
+  nntrainer::sharedConstTensor out = node->backwarding(deriv, iteration);
+
+  if (should_verify) {
+    verify(*out, expected_dx, err_msg);
+    verifyGrad(err_msg);
+  }
+
+  auto opt = node->getOptimizer();
+  if (opt) {
+    opt->apply_gradients(node->getWeights(), node->getNumWeights(), iteration);
+  }
+
+  return out;
+}
+
+GraphWatcher::GraphWatcher(const std::string &config) {
+  nn = nntrainer::NeuralNetwork();
+  nn.loadFromConfig(config);
+  nn.init();
+
+  FlatGraphType graph = nn.getFlatGraph();
+
+  for (auto it = graph.begin(); it != graph.end() - 1; ++it) {
+    nodes.push_back(NodeWatcher(*it));
+  }
+
+  loss_node = NodeWatcher(graph.back());
+}
+
+void GraphWatcher::compareFor(const std::string &reference,
+                              const nntrainer::TensorDim &label_shape,
+                              unsigned int iterations) {
+  std::ifstream ref(reference, std::ios_base::in | std::ios_base::binary);
+
+  if (ref.bad()) {
+    throw std::runtime_error("ref is bad!");
+  }
+
+  nntrainer::Tensor in(nn.getInputDimension());
+  nntrainer::Tensor lb(label_shape);
+
+  in.read(ref);
+  lb.read(ref);
+
+  auto prepareInitialWeight = [this]() {
+    std::for_each(nodes.begin(), nodes.end(),
+                  [](NodeWatcher &n) { n.cloneWeightsFromExpected(); });
+  };
+
+  auto matchWeightAfterUpdation = [this]() {
+    std::for_each(nodes.begin(), nodes.end(), [](NodeWatcher &n) {
+      n.verifyWeight("weight is diffrent after updation, check optimizer");
+    });
+  };
+
+  for (unsigned int iteration = 1; iteration <= iterations; ++iteration) {
+    nntrainer::sharedConstTensor input = MAKE_SHARED_TENSOR(in.clone());
+    nntrainer::sharedConstTensor label = MAKE_SHARED_TENSOR(lb.clone());
+
+    readIteration(ref);
+
+    iteration == 1 ? prepareInitialWeight() : matchWeightAfterUpdation();
```

Comment:
move `matchWeightAfterUpdation` after line 317.
move `prepareInitialWeight()` before this loop.
Then line 320 can also be removed. It makes easier to understand and reason whats happening here.

Reply:
We've discussed offline to manage this in the upcoming PR's as this need change to the golden data format